core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java

```diff
@@ -16,45 +16,84 @@
 
 package com.linecorp.armeria.server;
 
+import static com.google.common.base.Preconditions.checkArgument;
 import static com.linecorp.armeria.server.ServerBuilder.decorate;
 import static java.util.Objects.requireNonNull;
 
+import java.util.Set;
+import java.util.function.Consumer;
 import java.util.function.Function;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
 
 import com.linecorp.armeria.internal.server.RouteDecoratingService;
+import com.linecorp.armeria.internal.server.RouteUtil;
 import com.linecorp.armeria.internal.server.annotation.AnnotatedServiceExtensions;
 import com.linecorp.armeria.server.annotation.ExceptionHandlerFunction;
 import com.linecorp.armeria.server.annotation.RequestConverterFunction;
 import com.linecorp.armeria.server.annotation.ResponseConverterFunction;
 
 /**
- * Builds {@link ServiceConfig}s for a {@link VirtualHostBuilder}. All {@link ServiceConfig}s
- * built by this builder will be served under a context path.
+ * Builds {@link ServiceConfig}s for a {@link ServerBuilder} or {@link VirtualHostBuilder}.
+ * All {@link ServiceConfig}s built by this builder will be served under the specified context paths.
+ *
+ * <pre>{@code
+ * Server.builder()
+ *       .contextPath("/v1", "/v2")
+ *       .service(myService) // served under "/v1" and "/v2"
+ *       .and()
+ *       .virtualHost("foo.com")
+ *       .contextPath("/v3")
+ *       .service(myService) // served by virtual host "foo.com" under "/v3"
+ * }</pre>
  *
  * @param <T> the original type which will be returned once the {@link ServiceConfig}
  *            is built using {@link #and()}.
  */
-final class ContextPathServicesBuilder<T extends ServiceConfigsBuilder>
+public final class ContextPathServicesBuilder<T extends ServiceConfigsBuilder>
         implements ServiceConfigsBuilder {
 
+    private final Set<String> contextPaths;
     private final T parent;
     private final VirtualHostBuilder virtualHostBuilder;
 
-    ContextPathServicesBuilder(T parent, VirtualHostBuilder virtualHostBuilder) {
+    ContextPathServicesBuilder(T parent, VirtualHostBuilder virtualHostBuilder,
+                               String... contextPaths) {
+        checkArgument(contextPaths.length > 0, "At least one context path is required");
+        for (String contextPath: contextPaths) {
+            RouteUtil.ensureAbsolutePath(contextPath, "contextPath");
+        }
+
         this.parent = parent;
+        this.contextPaths = ImmutableSet.copyOf(contextPaths);
         this.virtualHostBuilder = virtualHostBuilder;
     }
 
+    /**
+     * Configures an {@link HttpService} under the context path with the {@code customizer}.
+     */
+    public ContextPathServicesBuilder<T> withRoute(
```

Comment:
Note that this method hasn't/can't be extract to the interface without a breaking change

Reply:
Do you mean extracting to `ServiceConfigsBuilder`?

Reply:
Right